Fix unlift to unblock training IR + run_decomp on aliasing constants

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137162

When we populate unlifted graph module, we actually only "unlift" constant tensor inputs which is problematic because export de-duplicates aliasing constants. As a result, we only register one constant instead of two constants. This PR fixes that by querying ep.constants table instead of ep.graph_signature.lifted_tensor_constants. 

Differential Revision: [D63743111](https://our.internmc.facebook.com/intern/diff/D63743111)

tugsbayasgalan committed:
Fix unlift

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63743111).

pianpwk commented on torch/export/_unlift.py:
```diff
@@ -298,10 +294,20 @@ def _create_stateful_graph_module(
 
+    # Constants are not preserved well when we create a new GraphModule unlike param/buffers
+    for const_name, value in ep.constants.items():
+        if isinstance(value, torch.Tensor):
```
should we have a check that this attribute is None, to make it clear this is only needed for aliased constants?

tugsbayasgalan replied:
Hmm you mean attribute is set None? Yeah i can.

pianpwk commented on torch/export/_unlift.py:
```diff
@@ -258,7 +254,7 @@ def _create_stateful_graph_module(
     range_constraints,
     # TODO(suo) this should not be optional, but is since we still ahve
     # capture_pre_autograd_graph grr
```
does this comment prevent us from unifying range_constrants & ep as one argument lol

tugsbayasgalan replied:
hahaah yes.

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63743111).

tugsbayasgalan committed:
Update on "Fix unlift to unblock training IR + run_decomp on aliasing constants"


When we populate unlifted graph module, we actually only "unlift" constant tensor inputs which is problematic because export de-duplicates aliasing constants. As a result, we only register one constant instead of two constants. This PR fixes that by querying ep.constants table instead of ep.graph_signature.lifted_tensor_constants. 

Differential Revision: [D63743111](https://our.internmc.facebook.com/intern/diff/D63743111)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63743111).

angelayi commented on torch/fx/graph_module.py:
```diff
@@ -28,6 +28,7 @@
 
 _USER_PRESERVED_ATTRIBUTES_KEY = "_user_preserved_attributes"
 
+
```
can you remove all the formatting changes?

angelayi commented on torch/export/_unlift.py:
```diff
@@ -298,10 +301,21 @@ def _create_stateful_graph_module(
 
+    # Constants are not preserved well when we create a new GraphModule unlike param/buffers
+    for const_name, value in ep.constants.items():
+        if not torch.fx.graph_module._has_attr(stateful_gm, const_name):
```
I think `_recursive_getattr` does the same thing?

tugsbayasgalan replied:
Yeah i will delete that and use the one in fx.graph_module. I think we have many duplicate implementation of a same thing haha.

tugsbayasgalan committed:
Update on "Fix unlift to unblock training IR + run_decomp on aliasing constants"


When we populate unlifted graph module, we actually only "unlift" constant tensor inputs which is problematic because export de-duplicates aliasing constants. As a result, we only register one constant instead of two constants. This PR fixes that by querying ep.constants table instead of ep.graph_signature.lifted_tensor_constants. 

Differential Revision: [D63743111](https://our.internmc.facebook.com/intern/diff/D63743111)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63743111).

tugsbayasgalan committed:
Update on "Fix unlift to unblock training IR + run_decomp on aliasing constants"


When we populate unlifted graph module, we actually only "unlift" constant tensor inputs which is problematic because export de-duplicates aliasing constants. As a result, we only register one constant instead of two constants. This PR fixes that by querying ep.constants table instead of ep.graph_signature.lifted_tensor_constants. 

Differential Revision: [D63743111](https://our.internmc.facebook.com/intern/diff/D63743111)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63743111).

tugsbayasgalan committed:
Update on "Fix unlift to unblock training IR + run_decomp on aliasing constants"


When we populate unlifted graph module, we actually only "unlift" constant tensor inputs which is problematic because export de-duplicates aliasing constants. As a result, we only register one constant instead of two constants. This PR fixes that by querying ep.constants table instead of ep.graph_signature.lifted_tensor_constants. 

Differential Revision: [D63743111](https://our.internmc.facebook.com/intern/diff/D63743111)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63743111).

jovianjaison commented:
@pytorchbot revert -m 'Sorry for reverting your changes but many jobs are failing with NameError: name '_recursive_getattr' is not defined + a Lint job fails' -c ignoredsignal

eg.
[GH job link](https://github.com/pytorch/pytorch/actions/runs/11166863715/job/31042201711)
[GH job link](https://github.com/pytorch/pytorch/actions/runs/11166863715/job/31042201973)
[GH job link](https://github.com/pytorch/pytorch/actions/runs/11166742594/job/31041389677)
[HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/f96020c246aec8514b945d530879635a03294f70)
